plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/MigrateProcedure.java

```diff
@@ -182,28 +206,150 @@ public void migrate(ConnectorSession session, String schemaName, String tableNam
+        }
+
+        ProtocolEntry protocolEntry = transactionLogAccess.getProtocolEntries(tableSnapshot, session)
+                .reduce((first, second) -> second)
+                .orElseThrow(() -> new TrinoException(DELTA_LAKE_INVALID_SCHEMA, "Protocol entry not found in transaction log for table " + tableSnapshot.getTable()));
+        MetadataEntry metadataEntry = transactionLogAccess.getMetadataEntry(tableSnapshot, session);
+
+        if (protocolEntry.getMinReaderVersion() > 1) {
+            throw new TrinoException(INVALID_FUNCTION_ARGUMENT, "Migration of Delta Lake tables using reader versions above 2 is not supported");
+        }
+        if (protocolEntry.getMinWriterVersion() > 2) {
+            throw new TrinoException(INVALID_FUNCTION_ARGUMENT, "Migration of Delta Lake tables using writer versions above 2 is not supported");
+        }
+
+        List<DeltaLakeColumnMetadata> deltaLakeColumns = DeltaLakeSchemaSupport.extractSchema(metadataEntry, typeManager);
+        AtomicInteger nextFieldId = new AtomicInteger(1);
+        List<Types.NestedField> icebergColumns = new ArrayList<>();
+        for (DeltaLakeColumnMetadata column : deltaLakeColumns) {
+            int index = icebergColumns.size();
+            org.apache.iceberg.types.Type type = toIcebergTypeForNewColumn(typeManager.getType(column.getType().getTypeSignature()), nextFieldId);
+            Types.NestedField field = Types.NestedField.of(index, column.getColumnMetadata().isNullable(), column.getName(), type, column.getColumnMetadata().getComment());
+            icebergColumns.add(field);
+        }
+        org.apache.iceberg.types.Type icebergSchema = Types.StructType.of(icebergColumns);
+        icebergSchema = TypeUtil.assignFreshIds(icebergSchema, nextFieldId::getAndIncrement);
+        Schema schema = new Schema(icebergSchema.asStructType().fields());
+
+        NameMapping nameMapping = MappingUtil.create(schema);
+        Map<String, String> properties = icebergTableProperties(location, hiveTable.getParameters(), nameMapping);
+        PartitionSpec partitionSpec = partitionSpecFromDeltaLake(schema, metadataEntry);
+
+        Map<String, DeltaLakeColumnHandle> columnHandles = deltaLakeColumns.stream().collect(toImmutableMap(
+                DeltaLakeColumnMetadata::getName,
+                columnMetadata -> new DeltaLakeColumnHandle(
+                        columnMetadata.getName(),
+                        columnMetadata.getType(),
+                        OptionalInt.empty(),
+                        columnMetadata.getPhysicalName(),
+                        columnMetadata.getPhysicalColumnType(),
+                        DeltaLakeColumnType.REGULAR)));
+        try {
+            ImmutableList.Builder<DataFile> dataFilesBuilder = ImmutableList.builder();
+            for (AddFileEntry addFileEntry : transactionLogAccess.getActiveFiles(tableSnapshot, session)) {
```

Comment:
This is performing the migration in a single thread right? Is it worthwhile to introduce parallelism, similar to the approach in this PR: https://github.com/trinodb/trino/pull/16786/files

Reply:
I'm becoming less sure we want to do that though. Re-reading the data file will give us fully populated Metrics for each file. Delta Lake has fewer file stats than Iceberg does so if we try to re-use what they have we'll be missing `columnSizes`, `valueCounts`, and `nanValueCounts`.

@ebyhr @findepi any thoughts on if ti's worth the I/O to load those from the data files?

Reply:
> have we'll be missing columnSizes, valueCounts, and nanValueCounts.

are these present in a file footer for files created by OSS Delta or DBR?


>  Delta tables should be much less heavy to migrate but it might still be a good idea.

maybe we should migrate using a table procedure

Reply:
> maybe we should migrate using a table procedure

I suggested the same in Vikash's PR adding [thread level parallelism](https://github.com/trinodb/trino/pull/16786#issuecomment-1511599060). I think it'd be good to do as a follow-up.

> are these present in a file footer for files created by OSS Delta or DBR?

Good question, I will check.

Reply:
> are these present in a file footer for files created by OSS Delta or DBR?

Testing against open source Delta Lake, they do write the Parquet metadata we'd need for `columnSizes` and `valueCounts`, but non `nanValueCounts`.